### PR TITLE
Add support for override request headers

### PR DESCRIPTION
**Pull request by nasundar**
1. Support overriding request headers
2. Support for logging gzip and zlib error responses
3. Log requestId in traces

### CODE REVIEW

File: src/contracts/encoder_decoder_contract.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package contracts
+
+import (
+	"bytes"
+	"io"
+)
+
+// Contract to encode and decode response body.
+type IEncodeDecode interface {
```
Comment (ms-meya) [inline src/contracts/encoder_decoder_contract.go:9]: nit: IEncoderDecoder

File: src/utils/configuration.go [hunk 2]
```diff
@@ -47,6 +50,11 @@ type telemetryConfiguration struct {
 	otelServiceName string
 }
 
+type additionalHeaders struct {
+	headers    map[string]string
+	headersStr string
```
Comment (ms-meya) [inline src/utils/configuration.go:55]: is this required? it seems like not being used? If not, they we don't need additionalHeaders struct

File: deploy/chart/aad-auth-proxy/values-template.yaml [hunk 3]
```diff
@@ -12,6 +12,7 @@ aadClientCertificatePath: ""
 aadTokenRefreshIntervalInPercentage: "10"
 otelServiceName: "aad_auth_proxy"
 otelGrpcEndpoint: ""
+overrideRequestHeaders: "{}"
```
Comment (ms-meya) [inline deploy/chart/aad-auth-proxy/values-template.yaml:15]: just curious, is this how a map chart value get configured? Did you consider something like below:
```
overrideRequestHeaders:
- name: "header1"
  value: "value1"
- name: "header2"
  value: "value2;value3"
```

Comment (nasundar) [reply deploy/chart/aad-auth-proxy/values-template.yaml:15]: This is one way of doing it, I just picked the simpler approach. The one you mentioned might work well with yaml based configuration in helm, not sure if I can do that in yaml for images, where I expect values to be env variables. Also, I thought keeping it generic will work for compiled code running on command line as well.

File: src/utils/configuration.go (hunk 2 above)
Comment (ms-meya) [reply src/utils/configuration.go:55]: sorry, I was meant to say `headerStr` seems not being used of this struct

Comment (nasundar) [reply src/utils/configuration.go:55]: I see, its being used to log value to console once we have parsed overrideHeaders

image

**Review by ms-meya — APPROVED** [review]

**Pull request merged**